### ISSUE TITLE
[RFC] TensorImpl::has_symbolic_sizes_strides can return false when sizes() actually contains symbolic values

### ISSUE THREAD

**Issue by davidberard98** [pytorch/pytorch#123698]
### 🐛 Describe the bug

**TL;DR**: I want to be able to call a function like `TensorImpl::may_have_symbolic_sizes_strides()` which will never return false if the tensor contains symbolic sizes / strides. Today, we have `TensorImpl::has_symbolic_sizes_strides()` which doesn't obey this property - in the repro at the bottom of this issue, we have an example of a Nested Tensor with symbolic sizes for which `has_symbolic_sizes_strides()` returns False. **My question - what's the right way to provide this API?**

**Background: has_symbolic_sizes_strides vs python-implemented sizes**:

has_symbolic_sizes_strides is set to true if, during construction of a TensorImpl, symbolic sizes are passed in:

https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/c10/core/TensorImpl.cpp#L852-L866

But in the case of tensor subclasses, we can provide the option to implement sizes/strides in python. Nested Tensor uses this option, in the 2nd input to `_make_wrapper_subclass`. Additionally, the 11th input is `"sizes"`, which implies something (described below)

https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/torch/nested/_internal/nested_tensor.py#L64-L74

The 11th input of "sizes" is the `dispatch_sizes_strides_policy`:

https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/torch/csrc/autograd/python_variable.cpp#L753

we use it to set_python_custom_sizes_strides; in this case, "sizes" gets parsed to `c10::TensorImpl::SizesStridesPolicy::CustomSizes`

https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/torch/csrc/autograd/python_variable.cpp#L753

following this pointer to set_python_custom_sizes_strides:

https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/c10/core/TensorImpl.h#L2779-L2782

and then to refresh_sizes_strides_policy:

https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/c10/core/TensorImpl.h#L2805-L2813

we see that this controls the sizes_strides_policy_, and also has handling based on has_symbolic_sizes_strides_. Note that `sizes_strides_policy_` and `has_symbolic_sizes_strides_` are separate

Meanwhile, Nested Tensor implements sizes via torch dispatch:

[TODO... more details, link to nt implementation]

**Ideas**

1.

**Repro**:
```python
import torch

values = torch.rand((8, 4))
offsets = torch.tensor([0, 2, 5, 6, 8])
nt = torch.nested.nested_tensor_from_jagged(values, offsets)
print(nt.shape)  # torch.Size([4, j2, 4])
print(nt._has_symbolic_sizes_strides)  # False
```

### Versions

main branch

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang

**Comment by soulitzer** [comment]
> I thought our agreement was that we were going to stop using dispatch to Python to implement sizes and just store directly in sizes field

This sounds like the right approach... I don't remember why we originally went the other route

**Comment by soulitzer** [comment]
Oh one reason to have dispatch to Python is that if we had a `[B, *, D]` tensor and we wanted its numel, we can directly use the numel of the values. It may be difficult to produce numel otherwise, since ordinarily we would try to multiply `B` with `*`.

**Comment by ezyang** [comment]
So, uh, don't compute numel that way? Lol

**Comment by soulitzer** [comment]
Hmmm yeah I guess having a SizesStridesPolicy specifically for NestedTensor would solve this, if that is what you mean

**Comment by davidberard98** [comment]
Temporary fix - https://github.com/pytorch/pytorch/pull/123696
But it sounds like we should still make the changes suggested by Ed and Jeffrey as a long-term fix.

**Comment by ezyang** [comment]
What I mean is, ordinarily, numel is computed off of sizes. But there's no reason it has to be; we maintain numel as its own field and we can populate it however we like.
